clippy_lints/src/box_default.rs

```diff
@@ -0,0 +1,58 @@
+use clippy_utils::{diagnostics::span_lint_and_help, is_default_equivalent, match_qpath, paths};
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// checks for `Box::new(T::default())`, which is better written as
+    /// `Box::<T>::default()`.
+    ///
+    /// ### Why is this bad?
+    /// First, it's more complex, involving two calls instead of one.
+    /// Second, `Box::default()` can be faster
+    /// [in certain cases](https://nnethercote.github.io/perf-book/standard-library-types.html#box).
+    ///
+    /// ### Known problems
+    /// The lint may miss some cases (e.g. Box::new(String::from(""))).
+    /// On the other hand, it will trigger on cases where the `default`
+    /// code comes from a macro that does something different based on
+    /// e.g. target operating system.
```

Comment:
Could add a check for `expr.span.ctxt() == arg.span.ctxt()` to avoid linting across macro boundaries while still catching it in local macro definitions

Reply:
Good idea. Done.

Reply:
Can remove the macro note from `Known problems` with that unless there's a case I'm missing

Reply:
Oh maybe things like `Box::new(cfg_if! ... { String::new() } else { ... })` where the expressions aren't coming from the macro

Reply:
Exactly. Though if you want, I can remove the note until I find a clearer wording.

Reply:
I'd say remove it, apparently you can't use `cfg_if` in an expression position, nor `#[cfg]` as attributes on expressions are unstable